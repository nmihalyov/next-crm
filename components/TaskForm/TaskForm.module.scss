.form {
  display: block;
  margin-bottom: 30px;
}

.row {
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
}

.field {
  width: calc(50% - 20px);
}