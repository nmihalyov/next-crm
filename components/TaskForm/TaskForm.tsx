import React from 'react';
import type { Task } from '../../types/task';
import useActions from '../../hooks/useActions';
import useInput from '../../hooks/useInput';
import useTypedSelector from '../../hooks/useTypedSelector';
import { Form, Input, Button } from 'antd';

const TaskForm: React.FC<{
  onApply(data: Task): void
}> = props => {
  const { onApply } = props;
  const input = useInput();
  const [form] = Form.useForm();
  const isApplying = useTypedSelector(state => state.tasks.isApplying);
  const { setApplyingTask } = useActions();

  const applyTask = (): void => {
    setApplyingTask(true);
    onApply({
      userId: 1,
      id: Date.now(),
      title: input.value,
      completed: false
    });
    input.clear();
    form.resetFields();
  };

  return (
    <Form
      autoComplete="off"
      form={form}
      labelCol={{span: 3}}
      onFinish={applyTask}>

      <Form.Item
        label="Task name"
        name="title"
        rules={[{required: true, message: 'Please, enter task title'}]}>
        <Input {...input.bind} name="title" allowClear />
      </Form.Item>

      <Form.Item wrapperCol={{offset: 3}}>
        <Button
          type="primary"
          htmlType="submit"
          size="large"
          loading={isApplying}>
          Apply task
        </Button>
      </Form.Item>

    </Form>
  );
};

export default React.memo(TaskForm);