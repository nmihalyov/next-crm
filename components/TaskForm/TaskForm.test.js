import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import TaskForm from './TaskForm';

describe('TaskForm:', () => {
  const onApply = jest.fn();
  const titleValue = 'Title';
  let component;
  let button;
  let input;
  let store;

  beforeAll(() => {
    const mockStore = configureStore([]);
    jest.spyOn(Date, 'now').mockImplementation(() => 1487076708000);

    store = mockStore({
      tasks: {
        data: [],
        isApplying: false,
        isFetching: false,
        isFetched: true
      }
    });
  });

  beforeEach(() => {
    onApply.mockRestore();
    component = mount(
      <Provider store={store}>
        <TaskForm onApply={onApply} />
      </Provider>
    );
    button = component.find('Button');
    input = component.find('input');
  });

  it('renders TaskForm', () => {
    expect(component).toMatchSnapshot();
  });

  // This is a dummy hack to run test after timeout, otherwise it fails (supposed to be two different tests)
  // If you have a better solution, please send pull-request or contact me directly
  it('calls onApply with correct arguments and doesn\'t call without at least one input not filled', done => {
    button.simulate('submit');

    setTimeout(() => {
      expect(onApply).not.toHaveBeenCalled();

      input.simulate('change', {target: {value: titleValue}});
      button.simulate('submit');

      setTimeout(() => {
        expect(onApply).toHaveBeenCalledWith({userId: 1, title: titleValue, completed: false, id: Date.now()});
        expect(input.prop('value')).toBe('');
        done();
      });
    });
  });
});