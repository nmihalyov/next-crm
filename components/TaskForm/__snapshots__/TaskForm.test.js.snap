// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskForm: renders TaskForm 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Memo(TaskForm)
    onApply={[MockFunction]}
  >
    <ForwardRef(InternalForm)
      autoComplete="off"
      form={
        Object {
          "__INTERNAL__": Object {
            "itemRef": [Function],
            "name": undefined,
          },
          "getFieldError": [Function],
          "getFieldInstance": [Function],
          "getFieldValue": [Function],
          "getFieldsError": [Function],
          "getFieldsValue": [Function],
          "getInternalHooks": [Function],
          "isFieldTouched": [Function],
          "isFieldValidating": [Function],
          "isFieldsTouched": [Function],
          "isFieldsValidating": [Function],
          "resetFields": [Function],
          "scrollToField": [Function],
          "setFields": [Function],
          "setFieldsValue": [Function],
          "submit": [Function],
          "validateFields": [Function],
        }
      }
      labelCol={
        Object {
          "span": 3,
        }
      }
      onFinish={[Function]}
    >
      <SizeContextProvider>
        <ForwardRef(Form)
          autoComplete="off"
          className="ant-form ant-form-horizontal"
          form={
            Object {
              "__INTERNAL__": Object {
                "itemRef": [Function],
                "name": undefined,
              },
              "getFieldError": [Function],
              "getFieldInstance": [Function],
              "getFieldValue": [Function],
              "getFieldsError": [Function],
              "getFieldsValue": [Function],
              "getInternalHooks": [Function],
              "isFieldTouched": [Function],
              "isFieldValidating": [Function],
              "isFieldsTouched": [Function],
              "isFieldsValidating": [Function],
              "resetFields": [Function],
              "scrollToField": [Function],
              "setFields": [Function],
              "setFieldsValue": [Function],
              "submit": [Function],
              "validateFields": [Function],
            }
          }
          onFinish={[Function]}
          onFinishFailed={[Function]}
        >
          <form
            autoComplete="off"
            className="ant-form ant-form-horizontal"
            onReset={[Function]}
            onSubmit={[Function]}
          >
            <FormItem
              label="Task name"
              name="title"
              rules={
                Array [
                  Object {
                    "message": "Please, enter task title",
                    "required": true,
                  },
                ]
              }
            >
              <WrapperField
                label="Task name"
                messageVariables={
                  Object {
                    "label": "Task name",
                  }
                }
                name="title"
                onReset={[Function]}
                rules={
                  Array [
                    Object {
                      "message": "Please, enter task title",
                      "required": true,
                    },
                  ]
                }
                trigger="onChange"
                validateTrigger="onChange"
              >
                <Field
                  fieldContext={
                    Object {
                      "__INTERNAL__": Object {
                        "itemRef": [Function],
                        "name": undefined,
                      },
                      "getFieldError": [Function],
                      "getFieldInstance": [Function],
                      "getFieldValue": [Function],
                      "getFieldsError": [Function],
                      "getFieldsValue": [Function],
                      "getInternalHooks": [Function],
                      "isFieldTouched": [Function],
                      "isFieldValidating": [Function],
                      "isFieldsTouched": [Function],
                      "isFieldsValidating": [Function],
                      "resetFields": [Function],
                      "scrollToField": [Function],
                      "setFields": [Function],
                      "setFieldsValue": [Function],
                      "submit": [Function],
                      "validateFields": [Function],
                      "validateTrigger": "onChange",
                    }
                  }
                  key="_title"
                  label="Task name"
                  messageVariables={
                    Object {
                      "label": "Task name",
                    }
                  }
                  name={
                    Array [
                      "title",
                    ]
                  }
                  onReset={[Function]}
                  rules={
                    Array [
                      Object {
                        "message": "Please, enter task title",
                        "required": true,
                      },
                    ]
                  }
                  trigger="onChange"
                  validateTrigger="onChange"
                  valuePropName="value"
                >
                  <Row
                    className="ant-form-item"
                    key="row"
                  >
                    <div
                      className="ant-row ant-form-item"
                      style={Object {}}
                    >
                      <FormItemLabel
                        htmlFor="title"
                        label="Task name"
                        name="title"
                        prefixCls="ant-form"
                        required={true}
                        requiredMark={true}
                        rules={
                          Array [
                            Object {
                              "message": "Please, enter task title",
                              "required": true,
                            },
                          ]
                        }
                      >
                        <Col
                          className="ant-form-item-label"
                          span={3}
                        >
                          <div
                            className="ant-col ant-col-3 ant-form-item-label"
                            style={Object {}}
                          >
                            <label
                              className="ant-form-item-required"
                              htmlFor="title"
                              title="Task name"
                            >
                              Task name
                            </label>
                          </div>
                        </Col>
                      </FormItemLabel>
                      <FormItemInput
                        errors={Array []}
                        label="Task name"
                        name={
                          Array [
                            "title",
                          ]
                        }
                        onDomErrorVisibleChange={[Function]}
                        prefixCls="ant-form"
                        rules={
                          Array [
                            Object {
                              "message": "Please, enter task title",
                              "required": true,
                            },
                          ]
                        }
                        status=""
                        touched={false}
                        validateStatus=""
                        validating={false}
                      >
                        <Col
                          className="ant-form-item-control"
                        >
                          <div
                            className="ant-col ant-form-item-control"
                            style={Object {}}
                          >
                            <div
                              className="ant-form-item-control-input"
                            >
                              <div
                                className="ant-form-item-control-input-content"
                              >
                                <Component
                                  update={1}
                                >
                                  <Input
                                    allowClear={true}
                                    id="title"
                                    name="title"
                                    onChange={[Function]}
                                    type="text"
                                  >
                                    <ClearableLabeledInput
                                      allowClear={true}
                                      bordered={true}
                                      element={
                                        <input
                                          className="ant-input"
                                          id="title"
                                          name="title"
                                          onBlur={[Function]}
                                          onChange={[Function]}
                                          onFocus={[Function]}
                                          onKeyDown={[Function]}
                                          type="text"
                                        />
                                      }
                                      focused={false}
                                      handleReset={[Function]}
                                      id="title"
                                      inputType="input"
                                      name="title"
                                      onChange={[Function]}
                                      prefixCls="ant-input"
                                      triggerFocus={[Function]}
                                      type="text"
                                      value=""
                                    >
                                      <span
                                        className="ant-input-affix-wrapper"
                                        onMouseUp={[Function]}
                                      >
                                        <input
                                          className="ant-input"
                                          id="title"
                                          name="title"
                                          onBlur={[Function]}
                                          onChange={[Function]}
                                          onFocus={[Function]}
                                          onKeyDown={[Function]}
                                          style={null}
                                          type="text"
                                          value=""
                                        />
                                        <span
                                          className="ant-input-suffix"
                                        >
                                          <ForwardRef(CloseCircleFilled)
                                            className="ant-input-clear-icon-hidden ant-input-clear-icon"
                                            onClick={[Function]}
                                            onMouseDown={[Function]}
                                            role="button"
                                          >
                                            <AntdIcon
                                              className="ant-input-clear-icon-hidden ant-input-clear-icon"
                                              icon={
                                                Object {
                                                  "icon": Object {
                                                    "attrs": Object {
                                                      "focusable": "false",
                                                      "viewBox": "64 64 896 896",
                                                    },
                                                    "children": Array [
                                                      Object {
                                                        "attrs": Object {
                                                          "d": "M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z",
                                                        },
                                                        "tag": "path",
                                                      },
                                                    ],
                                                    "tag": "svg",
                                                  },
                                                  "name": "close-circle",
                                                  "theme": "filled",
                                                }
                                              }
                                              onClick={[Function]}
                                              onMouseDown={[Function]}
                                              role="button"
                                            >
                                              <span
                                                aria-label="close-circle"
                                                className="anticon anticon-close-circle ant-input-clear-icon-hidden ant-input-clear-icon"
                                                onClick={[Function]}
                                                onMouseDown={[Function]}
                                                role="button"
                                                tabIndex={-1}
                                              >
                                                <IconReact
                                                  icon={
                                                    Object {
                                                      "icon": Object {
                                                        "attrs": Object {
                                                          "focusable": "false",
                                                          "viewBox": "64 64 896 896",
                                                        },
                                                        "children": Array [
                                                          Object {
                                                            "attrs": Object {
                                                              "d": "M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z",
                                                            },
                                                            "tag": "path",
                                                          },
                                                        ],
                                                        "tag": "svg",
                                                      },
                                                      "name": "close-circle",
                                                      "theme": "filled",
                                                    }
                                                  }
                                                >
                                                  <svg
                                                    aria-hidden="true"
                                                    data-icon="close-circle"
                                                    fill="currentColor"
                                                    focusable="false"
                                                    height="1em"
                                                    key="svg-close-circle"
                                                    viewBox="64 64 896 896"
                                                    width="1em"
                                                  >
                                                    <path
                                                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                                                      key="svg-close-circle-svg-0"
                                                    />
                                                  </svg>
                                                </IconReact>
                                              </span>
                                            </AntdIcon>
                                          </ForwardRef(CloseCircleFilled)>
                                        </span>
                                      </span>
                                    </ClearableLabeledInput>
                                  </Input>
                                </Component>
                              </div>
                            </div>
                            <ErrorList
                              errors={Array []}
                              onDomErrorVisibleChange={[Function]}
                            >
                              <CSSMotion
                                motionDeadline={500}
                                motionName="ant-show-help"
                                onLeaveEnd={[Function]}
                                visible={false}
                              >
                                <DomWrapper />
                              </CSSMotion>
                            </ErrorList>
                          </div>
                        </Col>
                      </FormItemInput>
                    </div>
                  </Row>
                </Field>
              </WrapperField>
            </FormItem>
            <FormItem
              wrapperCol={
                Object {
                  "offset": 3,
                }
              }
            >
              <Row
                className="ant-form-item"
                key="row"
              >
                <div
                  className="ant-row ant-form-item"
                  style={Object {}}
                >
                  <FormItemLabel
                    prefixCls="ant-form"
                    requiredMark={true}
                    wrapperCol={
                      Object {
                        "offset": 3,
                      }
                    }
                  />
                  <FormItemInput
                    errors={Array []}
                    onDomErrorVisibleChange={[Function]}
                    prefixCls="ant-form"
                    status=""
                    validateStatus=""
                    wrapperCol={
                      Object {
                        "offset": 3,
                      }
                    }
                  >
                    <Col
                      className="ant-form-item-control"
                      offset={3}
                    >
                      <div
                        className="ant-col ant-col-offset-3 ant-form-item-control"
                        style={Object {}}
                      >
                        <div
                          className="ant-form-item-control-input"
                        >
                          <div
                            className="ant-form-item-control-input-content"
                          >
                            <Button
                              htmlType="submit"
                              loading={false}
                              size="large"
                              type="primary"
                            >
                              <Wave>
                                <button
                                  className="ant-btn ant-btn-primary ant-btn-lg"
                                  onClick={[Function]}
                                  type="submit"
                                >
                                  <LoadingIcon
                                    existIcon={false}
                                    loading={false}
                                    prefixCls="ant-btn"
                                  >
                                    <CSSMotion
                                      motionName="ant-btn-loading-icon-motion"
                                      onAppearActive={[Function]}
                                      onAppearStart={[Function]}
                                      onEnterActive={[Function]}
                                      onEnterStart={[Function]}
                                      onLeaveActive={[Function]}
                                      onLeaveStart={[Function]}
                                      removeOnLeave={true}
                                      visible={false}
                                    >
                                      <DomWrapper />
                                    </CSSMotion>
                                  </LoadingIcon>
                                  <span
                                    key=".0"
                                  >
                                    Apply task
                                  </span>
                                </button>
                              </Wave>
                            </Button>
                          </div>
                        </div>
                        <ErrorList
                          errors={Array []}
                          onDomErrorVisibleChange={[Function]}
                        >
                          <CSSMotion
                            motionDeadline={500}
                            motionName="ant-show-help"
                            onLeaveEnd={[Function]}
                            visible={false}
                          >
                            <DomWrapper />
                          </CSSMotion>
                        </ErrorList>
                      </div>
                    </Col>
                  </FormItemInput>
                </div>
              </Row>
            </FormItem>
          </form>
        </ForwardRef(Form)>
      </SizeContextProvider>
    </ForwardRef(InternalForm)>
  </Memo(TaskForm)>
</Provider>
`;
